### PR TITLE
add rbac check for multi-tenancy

### PR DESCRIPTION
**Pull request by pdgetrf**
this check has two parts

1. verify that user tenant from x509 cert matches the requested resource tenant (from kubectl config). 
2. verify user's tenant is authorized by the rolebindings from the same tenant. 

both check skips if user belongs to the system tenant.

local test:

when user tenant from cert does not match the requested resource tenant
```bash
root@ip-172-31-27-32:~# kubectl --context=pengdu-context get pods
Error from server (Forbidden): pods is forbidden: User "pengdu" cannot list resource "pods" in API group "" in the namespace "pengdunamespace"
```

when user tenant matches cert and rolebinding tenant and is authorized
```bash
root@ip-172-31-27-32:~# kubectl --context=pengdu-context get pods
No resources found.
```

when user tenant matches cert and rolebinding tenant but is not authorized
```bash
root@ip-172-31-27-32:~# kubectl --context=pengdu-context create -f vanilla.yaml
Error from server (Forbidden): unknown
```

### CODE REVIEW

**Comment by pdgetrf** [comment]
fixing unit tests

File: pkg/registry/rbac/validation/rule.go [hunk 1]
```diff
@@ -212,6 +216,14 @@ func (r *DefaultRuleResolver) VisitRulesFor(user user.Info, namespace string, vi
 		} else {
 			sourceDescriber := &roleBindingDescriber{}
 			for _, roleBinding := range roleBindings {
+
+				// verify user tenant matches rolebinding tenant
+				// "to remove check for default later when rolebinding can
```
Comment (chenqianfzh) [inline pkg/registry/rbac/validation/rule.go]: Need to make changes to to get the rolebindings only for the given userTenant. We may have tens of thousands of userTenants, it is very inefficient if we check all the role bindings for all the tenants.

File: pkg/registry/rbac/validation/rule.go [hunk 2]
```diff
@@ -176,11 +177,14 @@ func (d *roleBindingDescriber) String() string {
...
 func (r *DefaultRuleResolver) VisitRulesFor(user user.Info, namespace string, visitor func(source fmt.Stringer, rule *rbacv1.PolicyRule, err error) bool) {
+
+	userTenant := user.GetTenant()
+
 	if clusterRoleBindings, err := r.clusterRoleBindingLister.ListClusterRoleBindings(); err != nil {
 		if !visitor(nil, nil, err) {
 			return
 		}
-	} else {
+	} else if userTenant == "system" {
```
Comment (chenqianfzh) [inline pkg/registry/rbac/validation/rule.go]: Minor: we have a const definition, metav1.TenantSystem = "system"

Also, we have metav1.TenantDefault = "default"

File: pkg/registry/rbac/validation/rule.go [hunk 3]
```diff
@@ -176,11 +177,14 @@ func (d *roleBindingDescriber) String() string {
 }
 
 func (r *DefaultRuleResolver) VisitRulesFor(user user.Info, namespace string, visitor func(source fmt.Stringer, rule *rbacv1.PolicyRule, err error) bool) {
+
+	userTenant := user.GetTenant()
+
 	if clusterRoleBindings, err := r.clusterRoleBindingLister.ListClusterRoleBindings(); err != nil {
```
Comment (chenqianfzh) [inline pkg/registry/rbac/validation/rule.go]: As we discussed before, the ClusterRole and ClusterRoleBinding will be changed to tenant-scope. I will send out a PR on that. And you can make changes to get the rolebindings for the specific tenant only.

File: pkg/registry/rbac/validation/rule.go [hunk 4]
```diff
@@ -176,11 +177,14 @@ func (d *roleBindingDescriber) String() string {
 }
 
 func (r *DefaultRuleResolver) VisitRulesFor(user user.Info, namespace string, visitor func(source fmt.Stringer, rule *rbacv1.PolicyRule, err error) bool) {
+
+	userTenant := user.GetTenant()
```
Comment (chenqianfzh) [inline pkg/registry/rbac/validation/rule.go]: Do we need to handle the case that userTenant is empty?

File: staging/src/k8s.io/apiserver/pkg/authentication/request/union/union.go [hunk 5]
```diff
@@ -1,5 +1,6 @@
 /*
 Copyright 2014 The Kubernetes Authors.
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/union/union.go]: weird. Why the copyright statement is changed but there is not change in this file?

File: pkg/apis/authorization/types.go [hunk 6]
```diff
@@ -87,6 +87,8 @@ type ResourceAttributes struct {
 	Subresource string
 	// Name is the name of the resource being requested for a "get" or deleted for a "delete". "" (empty) means all.
 	Name string
+	// Tenant is the tenent of the resource being request
```
Comment (chenqianfzh) [inline pkg/apis/authorization/types.go]: typO: tenent

File: pkg/registry/rbac/validation/rule.go (hunk 4 above)
Comment (pdgetrf) [reply pkg/registry/rbac/validation/rule.go]: userTenant being empty sounds to me like coming into border checkpoint without a valid passport.

File: pkg/registry/rbac/validation/rule.go (hunk 2 above)
Comment (pdgetrf) [reply pkg/registry/rbac/validation/rule.go]: fixed

**Comment by pdgetrf** [comment]
will add more unit tests after fixing the current failing tests

File: pkg/registry/rbac/validation/rule.go (hunk 3 above)
Comment (chenqianfzh) [reply pkg/registry/rbac/validation/rule.go]: The change is now ready at https://github.com/futurewei-cloud/arktos/pull/117

File: staging/src/k8s.io/apiserver/pkg/authentication/request/union/union.go (hunk 5 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/union/union.go]: I had some changes but was later removed.

File: pkg/registry/rbac/validation/rule.go (hunk 3 above)
Comment (pdgetrf) [reply pkg/registry/rbac/validation/rule.go]: taken

File: pkg/registry/rbac/validation/rule.go (hunk 1 above)
Comment (pdgetrf) [reply pkg/registry/rbac/validation/rule.go]: your changes have been picked up for both rolebinding and clusterrolebinding. thanks!

**Comment by pdgetrf** [comment]
@chenqianfzh this is ready for review now. All your related changes have been merged.

File: api/openapi-spec/swagger.json [hunk 7]
```diff
@@ -19282,7 +19290,7 @@
   },
   "info": {
     "title": "Kubernetes",
-    "version": "v2.17.1"
```
Comment (chenqianfzh) [inline api/openapi-spec/swagger.json]: revert this version change.

Comment (pdgetrf) [reply api/openapi-spec/swagger.json]: didn't change this manually. I'll do another make update.

**Comment by pdgetrf** [comment]
hitting this error now in INT

Internal error occurred: failed to allocate a serviceIP: range is full\",\"reason\":\"InternalError\",\"details\":{\"causes\":[{\"message\":\"failed to allocate a serviceIP: range is full\"}]},\"code\":500}\n"

**Pull request merged**